sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainMessageProcessor.scala

```diff
@@ -0,0 +1,173 @@
+import io.horizen.params.NetworkParams
+import io.horizen.sc2sc.{CrossChainMessage, CrossChainMessageHash, CrossChainMessageImpl, CrossChainProtocolVersion}
+import io.horizen.utils.ZenCoinsUtils
+import org.web3j.abi.TypeReference
+import org.web3j.abi.datatypes.{StaticStruct, Type}
+import org.web3j.abi.datatypes.generated.Uint32
+import sparkz.crypto.hash.Keccak256
+
+import java.math.BigInteger
+import java.util
+import scala.collection.JavaConverters.seqAsJavaListConverter
+
+trait CrossChainMessageProvider {
+  private[horizen] def getCrossChainMessages(epochNum: Int, view: BaseAccountStateView): Seq[CrossChainMessage]
+  private[horizen] def getCrossChainMessageHashEpoch(msgHash: CrossChainMessageHash, view: BaseAccountStateView): Option[Int]
+}
+abstract class AbstractCrossChainMessageProcessor(networkParams: NetworkParams) extends NativeSmartContractMsgProcessor with CrossChainMessageProvider {
+
+  val MaxCrosschainMessagesPerEpoch = CryptoLibProvider.sc2scCircuitFunctions.getMaxMessagesPerCertificate
+  val DustThresholdInWei: BigInteger = ZenWeiConverter.convertZenniesToWei(ZenCoinsUtils.getMinDustThreshold(ZenCoinsUtils.MC_DEFAULT_FEE_RATE))
+
+  protected def execGetListOfCrosschainMessages(msg: Message, view: BaseAccountStateView): Array[Byte] = {
+    if (msg.getValue.signum() != 0) {
+      throw new ExecutionRevertedException("Call value must be zero")
+    }
+
+    if (msg.getData.length != METHOD_ID_LENGTH + GetListOfCrosschainMessagesCmdInputDecoder.getABIDataParamsLengthInBytes)
+      throw new ExecutionRevertedException(s"Wrong message data field length: ${msg.getData.length}")
+    val inputParams = GetListOfCrosschainMessagesCmdInputDecoder.decode(getArgumentsFromData(msg.getData))
+    val list = getListOfCrossChainMessagesRecords(inputParams.epochNum, view)
+    CrosschainMessagesListEncoder.encode(list.asJava)
+  }
+
+  override def getCrossChainMessages(epochNum: Int, view: BaseAccountStateView): Seq[CrossChainMessage] = {
+    getListOfCrossChainMessagesRecords(epochNum, view).map(msg => AbstractCrossChainMessageProcessor.buildCrosschainMessageFromAccount(msg, networkParams))
+  }
+
+  private[horizen] def getListOfCrossChainMessagesRecords(epochNum: Int, view: BaseAccountStateView): Seq[AccountCrossChainMessage] = {
+    val num: Int = getMessageEpochCounter(view, epochNum)
+
+    val list = (1 to num).map(index => {
+      val currentKey = getMessageKey(epochNum, index)
+      AccountCrossChainMessageSerializer.parseBytes(view.getAccountStorageBytes(contractAddress, currentKey))
+    })
+    list
+  }
+
+
+  override private[horizen] def getCrossChainMessageHashEpoch(messageHash: CrossChainMessageHash, view: BaseAccountStateView): Option[Int] = {
+    val data = view.getAccountStorage(contractAddress, messageHash.bytes)
+    if (data.sameElements(NULL_HEX_STRING_32)) {
+      Option.empty
+    } else {
+      Some(Ints.fromByteArray(data))
+    }
+  }
+
+
+  protected def addCrossChahinMessage(messageType: Int,
```

Comment:
```suggestion
  protected def addCrossChainMessage(messageType: Int,
```
Beside the name method may become final.
Same for some others methods, we for sure don't want storage related methods to be overridden. It may break the class.